### PR TITLE
Add wrapper class for loading permission info (Fix #12474)

### PR DESCRIPTION
**Pull request by joshmoore**
Since no solution for properly loading a Permissions object
when the target of a project query (`select d.details.permissions`),
this workaround is being added which will allow wrapping a
object in HQL (`select new ome.util.PermDetails(d)`) to produce
the desired effect.

#### Testing ####

Primary effect of this PR is that the test_permissions.py tests which were marked xfail should start passing:

```
components/tools/OmeroPy$ ./setup.py test -t test/integration/test_permissions.py -k Projection -svx
...
test/integration/test_permissions.py::TestPermissionProjections::testProjectionPermissions[rw-----admin-admin] PASSED
test/integration/test_permissions.py::TestPermissionProjections::testProjectionPermissions[rw-----admin-owner] PASSED
test/integration/test_permissions.py::TestPermissionProjections::testProjectionPermissionsWorkaround[rwrw---member-owner] PASSED
test/integration/test_permissions.py::TestPermissionProjections::testProjectionPermissionsWorkaround[rwrw---member-member] PASSED
test/integration/test_permissions.py::TestPermissionProjections::testExtendedRestrictions[PlateI] SKIPPED
test/integration/test_permissions.py::TestPermissionProjections::testExtendedRestrictions[WellI] SKIPPED
test/integration/test_permissions.py::TestPermissionProjections::testExtendedRestrictions[ImageI] SKIPPED
test/integration/test_permissions.py::TestPermissionProjections::testExtendedRestrictions[FileAnnotationI] SKIPPED

================================================================== 30 tests deselected by '-kProjection' ===================================================================
========================================================== 72 passed, 4 skipped, 30 deselected in 141.34 seconds ===========================================================
```

**For there to be further effects, queries will need to be changed from `select d.details.permissions` to `new map(d as d_details_permissions)`.**

### CODE REVIEW

**Comment by joshmoore** [comment]
cc: @will-moore @dpwrussell @jburel @chris-allan @knabar

**Comment by will-moore** [comment]
@joshmoore What is the correct way of doing something like this:
```
        select project.id,
               project.name,
               project.details.owner.id,
               new ome.util.PermDetails(project),
               (select count(id) from ProjectDatasetLink pdl
                where pdl.parent = project.id)
        from Project project
```
since this gives me a Query Exception:
```
unexpected token: , near line 4, column 40 [
```

**Comment by joshmoore** [comment]
I fear we may have just run into https://hibernate.atlassian.net/browse/HHH-2460

**Comment by will-moore** [comment]
Hmmm - bummer. Doesn't look like that is going to be fixed. Any ideas for a workaround?

**Comment by joshmoore** [comment]
@will-moore : can you give https://github.com/joshmoore/openmicroscopy/commit/7785939f9278c0c387ff09fbc90dd92f9bbdb039 a try? (We're now deeper into the HHH rabbit hole...)

**Comment by will-moore** [comment]
@joshmoore It's looking good... thanks!

**Comment by joshmoore** [comment]
NB: https://github.com/openmicroscopy/openmicroscopy/pull/3919 should be updated before merging.

**Comment by joshmoore** [comment]
Adding `breaing` to match gh-3995.

**Comment by joshmoore** [comment]
Removed breaking as per https://github.com/openmicroscopy/openmicroscopy/pull/3995#issuecomment-134636074

File: components/model/src/ome/util/PermDetails.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+/**
+ * wrapper class which can be used in HQL queries to properly load
+ * the full context for a permissions object. Rather than writing a
+ * query of the form:
+ *
+ * <code>
+ * select d.details.permissions from Dataset d
+ * </code>
+ *
+ * which returns a {@link Permissions} object with none of the extended
```
Comment (sbesson) [inline components/model/src/ome/util/PermDetails.java]: This link generates a Javadoc warning

```
Constructing Javadoc information...
Standard Doclet version 1.8.0_51
Building tree for all the packages and classes...
/opt/hudson/workspace/OMERO-DEV-merge-build/src/components/model/src/ome/util/PermDetails.java:54: warning - Tag @link: reference not found: Permissions
Building index for all the packages and classes...
```

File: components/model/src/ome/util/PermDetails.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+ * restrictions (canRead, canAnnotate, etc) properly loaded, use:
+ *
+ * <code>
+ * select new ome.util.PermDetails(d) from Dataset d
+ * </code>
+ *
+ * The return value for each will be the same.
+ *
+ * @see "HHH-3868"
+ * @see "trac-12474"
```
Comment (sbesson) [inline components/model/src/ome/util/PermDetails.java]: Use `<a href=URL>` for both links?

**Comment by joshmoore** [comment]
Pushed

**Comment by sbesson** [comment]
https://ci.openmicroscopy.org/view/DEV/job/OMERO-DEV-merge-integration-python/64/testReport/OmeroPy.test.integration.test_permissions/TestPermissionProjections/ certainly looks good as well as the wrapper class as a workaround. Deferring to the jstree testing team to know whether this has caused some performance issues /cc @pwalczysko?

**Comment by pwalczysko** [comment]
@joshmoore @sbesson : The last jstree testing was performed on 28 August . No performance issues were detected. Is this date including this PR ?

**Comment by sbesson** [comment]
@pwalczysko definitely, I think this PR was always included in the jstree testing because there is a dependency.

**Comment by joshmoore** [comment]
Definitely. This has been out of breaking for over a week (the entire time jstree has been out of breaking).

**Comment by pwalczysko** [comment]
In that case no objections.

**Comment by joshmoore** [comment]
--no-rebase

**Pull request merged**